### ISSUE TITLE
lxc copy: 

### ISSUE THREAD

**Issue by tomposmiko** [lxc/lxd#1827]
# Required information

 * Distribution: Ubuntu
 * Distribution version: Trusty
 * The output of "lxc info" or if that fails:
   * Kernel version: 4.2.0-34
   * LXC version: 2.0.0-rc13
   * LXD version: 2.0.0-rc7
   * Storage backend in use: zfs

# Issue description

Container cannot get copied.

# Steps to reproduce
```
# lxc copy zebra-ml:zebra-ml-ldapadmin users --debug
DBUG[03-30|13:28:21] Raw response: {"type":"sync","status":"Success","status_code":200,"metadata":{"architecture":"x86_64","config":{"volatile.base_image":"7c0b8de47a5b4353281b8224e9b3f092eb71de64d9014cb70114c408e6912b9d","volatile.eth0.hwaddr":"00:16:3e:36:01:fd","volatile.last_state.idmap":"[{\"Isuid\":true,\"Isgid\":false,\"Hostid\":231072,\"Nsid\":0,\"Maprange\":65536},{\"Isuid\":false,\"Isgid\":true,\"Hostid\":231072,\"Nsid\":0,\"Maprange\":65536}]"},"created_at":"2016-03-03T08:58:41Z","devices":{"root":{"path":"/","type":"disk"}},"ephemeral":false,"expanded_config":{"volatile.base_image":"7c0b8de47a5b4353281b8224e9b3f092eb71de64d9014cb70114c408e6912b9d","volatile.eth0.hwaddr":"00:16:3e:36:01:fd","volatile.last_state.idmap":"[{\"Isuid\":true,\"Isgid\":false,\"Hostid\":231072,\"Nsid\":0,\"Maprange\":65536},{\"Isuid\":false,\"Isgid\":true,\"Hostid\":231072,\"Nsid\":0,\"Maprange\":65536}]"},"expanded_devices":{"eth0":{"name":"eth0","nictype":"bridged","parent":"lxcbr0","type":"nic"},"root":{"path":"/","type":"disk"}},"name":"zebra-ml-ldapadmin","profiles":["default"],"stateful":false,"status":"Running","status_code":103}}
 
DBUG[03-30|13:28:21] Raw response: {"type":"sync","status":"Success","status_code":200,"metadata":{"api_extensions":[],"api_status":"development","api_version":"1.0","auth":"trusted","config":{"storage.zfs_pool_name":"tank/lxd"},"environment":{"addresses":[],"architectures":["x86_64","i686"],"certificate":"-----BEGIN CERTIFICATE-----\nMIIFuzCCA6OgAwIBAgIRANk+QpTjDB9H/vwVWDm8uPUwDQYJKoZIhvcNAQELBQAw\nQDEcMBoGA1UEChMTbGludXhjb250YWluZXJzLm9yZzEgMB4GA1UEAwwXcm9vdEB6\nZWJyYS5jaGVtYXhvbi5jb20wHhcNMTYwMzIxMTMyODExWhcNMjYwMzE5MTMyODEx\nWjBAMRwwGgYDVQQKExNsaW51eGNvbnRhaW5lcnMub3JnMSAwHgYDVQQDDBdyb290\nQHplYnJhLmNoZW1heG9uLmNvbTCCAiIwDQYJKoZIhvcNAQEBBQADggIPADCCAgoC\nggIBALPFpbB6fT6m9RC2Dboz+xTVbOk7bHSnU22zRbcNjsNIMEaVp2vvzw1S5Uo1\nM0BeUM9wiKqqpHjAB8Zod3h4Vp9id6ZRi8MH5Fcjo1L64JNT0L6mNU/CF6Tcn4rJ\nP9QHu9+AMMbpcinvhPN291m5DoN9xebDpiIgTUT/NTWV8DGYqB1Nr2GWoJ/bq4wB\nwq6rnbZLk24K+yor2qODWGVC+55zFlypxo6QNUGtH0V9b9ErwWHL2Q79kAKXfu/X\nSaA9zaUvhXcTalMOc2NvzL41ciVwizDFhSwFDUh2cdFjRk+aqytedQRH/ARs4DHU\nOTfGo0II11Qy/tkfBwU442UV5EZMtOeb6fr8jK5m57fPSh+vzYsc6hwpSQ6r7RXl\nR76+VKirmqi6l1zIoc1/yum2dGVlfeHtjdanM6RhF0uNP/s9eVt/FCN2GLXzB5us\nN6tvO3oC+lnxmBxsz3g+N4lLqameZi19Wwe3TX81AvqGFdIGit476yHb8mKjPQhU\n9pBryYH4G3EQRPKWuc/7yID+Kk/QcPsSCiBUXRo1anHrYz5H30dMUZXeryxw9OvF\nxD2NQG9BqMolSowt48+m+F61oQW5TVoaE8I+9zMGkRvau9paxa8NYh5ZnJleh3JA\n7p6OjPBjLuA+N1qiHPvmMnwBphhFBWsNIIUowzajNWL5L/fbAgMBAAGjga8wgaww\nDgYDVR0PAQH/BAQDAgWgMBMGA1UdJQQMMAoGCCsGAQUFBwMBMAwGA1UdEwEB/wQC\nMAAwdwYDVR0RBHAwboISemVicmEuY2hlbWF4b24uY29tghExODUuNDMuMjA2LjIx\nMC8yNIIaZmU4MDo6MjE2OjNjZmY6ZmViZjoyYTYvNjSCCzEwLjAuMy4xLzI0ghxm\nZTgwOjo4MDZhOjZmZmY6ZmU5ZjpmOThiLzY0MA0GCSqGSIb3DQEBCwUAA4ICAQB3\n0RJ1L8KXqVF5AonJ6lUMtoPOcvptz1VBLiMZNn3ZMP0Wb+1bgTm5Ue3N26nySgym\noaAkx9RIRDUNTTi5v2QQ72YZ1e6m4BBRE2UrHR1J5r6f5MVCSPDapRgA/DbBLbrr\npD88SkcPMt6vqZ7fjbWGu7iUQCRzMoahyPn64FpmlwKmld8h0Nl2yXWGuCgd3Wr1\nSHqHA6agqvfNRzXmsjBwC8c4Io9eEU89A35eWSruehvWFXH1SpvrTMDHppM0LDug\nCly8tsxA1T4y1goVz8miJM3NNxWptjsZO3XfqSKPw8DaKYwtDh9P8Em1GdjQ+BoR\np2GnftZl8O0pPp5wuffJy2A5NzdYnl2wVtlECyvdUS5G3GiO2mnrZjMrevCg7krl\ngik9zptg5Vq1ZIhNLvZ4tSQdcUUfoUvfZ+mfkIvWso/SY1bXMVgFY8P/rZzpxeRm\ncm4Fbde7YLs10E76VU52DGolp7n/R7ZbG9wkCGiRZeUCuEFXLz1WkJmMZ5Q+4Y+d\nR2k1Ymb3w6XIg7kIOXRsDSxN4YPtiNYI3xQE7dDLbwXG/UtJZwvChUw5i+mSqYGM\nzBYfzaiJSoykoNmhllKnNSZmX4i+jBIOa0iW0wfHpclewgcYKrcqBTaYvuQOAOcQ\nf+QkJpogZvtU2l7/jc8vb6n348nqawNVHRet/1DYlw==\n-----END CERTIFICATE-----\n","driver":"lxc","driver_version":"2.0.0.rc13","kernel":"Linux","kernel_architecture":"x86_64","kernel_version":"4.2.0-34-generic","server":"lxd","server_pid":31309,"server_version":"2.0.0.rc7","storage":"zfs","storage_version":"5"},"public":false}}
 
DBUG[03-30|13:28:21] Raw response: {"type":"sync","status":"Success","status_code":200,"metadata":["/1.0/profiles/default","/1.0/profiles/docker"]}
 
DBUG[03-30|13:28:21] Posting {"migration":true}
 to https://localhost:8443/1.0/containers/zebra-ml-ldapadmin 
DBUG[03-30|13:28:21] Raw response: {"type":"async","status":"Operation created","status_code":100,"metadata":{"id":"6bfb8008-f166-4502-9374-da396fa6f602","class":"websocket","created_at":"2016-03-30T12:28:21.947160954Z","updated_at":"2016-03-30T12:28:21.947160954Z","status":"Running","status_code":103,"resources":{"containers":["/1.0/containers/zebra-ml-ldapadmin"]},"metadata":{"control":"91080e7b403fbea18f9e29e873b3dc4ad65ced858fc9a4abb374ecf573761ab8","criu":"28c226cf9370f6e944968582c94d2ea277887a6379bdaf3c5de44c37a6afc66a","fs":"07c0f5d0c7937b548efbbbfa40e85f871605239b3cde6d5f147d42f3c36053e5"},"may_cancel":false,"err":""},"operation":"/1.0/operations/6bfb8008-f166-4502-9374-da396fa6f602"}
 
DBUG[03-30|13:28:21] Posting {"architecture":"x86_64","config":{},"devices":{"root":{"path":"/","type":"disk"}},"ephemeral":false,"name":"users","profiles":["default"],"source":{"base-image":"7c0b8de47a5b4353281b8224e9b3f092eb71de64d9014cb70114c408e6912b9d","certificate":"-----BEGIN CERTIFICATE-----\nMIIFXDCCA0SgAwIBAgIQOgP6Yw4uR3BNMDTDxJXWEjANBgkqhkiG9w0BAQsFADA2\nMRwwGgYDVQQKExNsaW51eGNvbnRhaW5lcnMub3JnMRYwFAYDVQQDDA1yb290QHpl\nYnJhLW1sMB4XDTE2MDIyNjEzMDIwMloXDTI2MDIyMzEzMDIwMlowNjEcMBoGA1UE\nChMTbGludXhjb250YWluZXJzLm9yZzEWMBQGA1UEAwwNcm9vdEB6ZWJyYS1tbDCC\nAiIwDQYJKoZIhvcNAQEBBQADggIPADCCAgoCggIBAKKaS1Im1+UcLhBn54xTEqro\n6C8g+0MV1sM3gOnp6P3iPg8PERy+hHR/Y7svw+IACago+qSOcaFXmSbQgsk8ogU8\ntzXT0G7w1bm1E59wTk5Hk+fvPyx6LYffK+M4WVA9EvjN2cANCN///ehTwApYoMtJ\nuVUUyV3yCGuzHWFeIAUV68M+d1Rtu5bDo+Id49PWn7tTbXfR2kmMc7Lbd3v5tbdw\nOMALS7I3sxJim0iZd4j3hafy6AefbPmmgKnUBuZ1u8I8n+Z5mK4J6AQlh5J0FszJ\nQzLzkedZ4QjGZDgXDWwEe4FhNoOnL9iJmCAbnLO8KqJ+8O+UPAv8rXccuyeYoWMw\nOXsnFLmXRIfmT8thmms8GNO1e8x0e4PU+nCCxpmapKlBO9p8ZHsKfn1p5m0Z4uXo\nMOPzKHvDCNKIkpyU76LXtHzPFElOKEdakvNuJR6dAeHIXpNnB/2W1/jXFnO7zuS4\nk/vgcPVTs24ZeDcBtPFaR5SzUCD718qOUAZjDV+CiAks6+HV351twVz1RYMPG/c4\n6g1FvaTXJC7GmHULauSy6obbdWkw8cTdjEmuJBniofhBj+RUwDbU61FEatYnzqiK\nnSxq/UVA7e82lq4YKMtmxzzAfFu5gwFwrJRKpK38HUioo1F0zSyn07j65RI+C+jl\nMi7qwh/E0vqVawpFr3d1AgMBAAGjZjBkMA4GA1UdDwEB/wQEAwIFoDATBgNVHSUE\nDDAKBggrBgEFBQcDATAMBgNVHRMBAf8EAjAAMC8GA1UdEQQoMCaCCHplYnJhLW1s\ngg0xMC4yNDAuMC42LzMyggsxMC4wLjMuMS8yNDANBgkqhkiG9w0BAQsFAAOCAgEA\nUQBXYAZWhyBDuJXRo8JbaI3XuCAk8LFXOCVAtEe0pEpd6byKNSHgHr4NHMixLwXh\n6DXOfbIY7zIqDJP4T7trXPk4e9/ySPjjlg+fseVZBPGx0W/8tfTMchct2HzeStsY\nRH5QCBkQjjpT9bdZcIUfMArOWy3SRZ+foISJPVWLG2MWU2X2vg3U8P9zL/FK6uyG\nqL9eN5mvVIRROCO2qeOacoj8734qaKnG6LwF9SGOoio0Nk/jGwTcsyztHtOVaQV+\n+kcZGAQXlOSGURoBr5osW8pYPeHWIDDY4QFWoUgHpLyw0hCrU1Cwgp2qVCddDncT\n4C9uOaQKaWyRjRuVi/cp6+N6CrZU9xR7Td6tprT8jvp84uBWqfiaW6d6lB4O3+N6\ntejLM1gQjRqzj06zlLPiRsQ48Lx+ClWVOOkAF+0IyAD6DezVZ9TwScqR2v7FZ6/i\nGqXehwyMySt39Bp/XEcdApkKfGmzDmj28adSAz0I3LaoeESvxfrcqkAcjT3Pki6a\nqh9A7mmsBm7QPfPqoV3Rss5R86q0ew1fL0aVFvKj3IAX2t7okbJc8XNNn04u3b1/\n+arFyM5oXL3Fxj4uVzxP7vNbNWI4z/pvRqc7hlF54dUIdTxa+8/PzvSJIyQH7HnL\nvKhMf7DA2EwUWeZdgPTS7QDTNiYz7+m9iajdOdTOgKY=\n-----END CERTIFICATE-----\n","mode":"pull","operation":"https://localhost:8443/1.0/operations/6bfb8008-f166-4502-9374-da396fa6f602","secrets":{"control":"91080e7b403fbea18f9e29e873b3dc4ad65ced858fc9a4abb374ecf573761ab8","criu":"28c226cf9370f6e944968582c94d2ea277887a6379bdaf3c5de44c37a6afc66a","fs":"07c0f5d0c7937b548efbbbfa40e85f871605239b3cde6d5f147d42f3c36053e5"},"type":"migration"}}
 to http://unix.socket/1.0/containers 
DBUG[03-30|13:28:21] Raw response: {"type":"async","status":"Operation created","status_code":100,"metadata":{"id":"0c4f6a66-1c46-4a12-ab0c-1e3c6dd0d3ca","class":"task","created_at":"2016-03-30T13:28:21.964487525+01:00","updated_at":"2016-03-30T13:28:21.964487525+01:00","status":"Running","status_code":103,"resources":{"containers":["/1.0/containers/users"]},"metadata":null,"may_cancel":false,"err":""},"operation":"/1.0/operations/0c4f6a66-1c46-4a12-ab0c-1e3c6dd0d3ca"}
 
DBUG[03-30|13:28:21] 1.0/operations/0c4f6a66-1c46-4a12-ab0c-1e3c6dd0d3ca/wait 
DBUG[03-30|13:28:23] Raw response: {"type":"sync","status":"Success","status_code":200,"metadata":{"id":"0c4f6a66-1c46-4a12-ab0c-1e3c6dd0d3ca","class":"task","created_at":"2016-03-30T13:28:21.964487525+01:00","updated_at":"2016-03-30T13:28:21.964487525+01:00","status":"Failure","status_code":400,"resources":{"containers":["/1.0/containers/users"]},"metadata":null,"may_cancel":false,"err":"Error transferring container data: sql: no rows in result set"}}
```

**Comment by stgraber** [comment]
The server logs of both hosts would be useful (/var/lib/lxd/lxd.log).

**Comment by stgraber** [comment]
I just tried copying a bunch of containers around on zfs and it all worked fine, so marking incomplete until we get more logs.

**Comment by stgraber** [comment]
Just did a read through the code and it's near impossible to figure out where that error came from.

So debug logs of both hosts would help, that or a reliable reproducer since normal container handling works fine here (and in our testsuite).

**Comment by stgraber** [comment]
I think I found the source of this, it's happening when the source container has extra zfs snapshots. Those snapshots are created when copying a container snapshot into a new container across the network.

I've got a fix which should fix the cause of this. Unfortunately there's not much we can do about existing containers as we can't distinguish the accidental snapshots from real ones.

Your best bet when that happens is to do a "zfs list" on the source host and look for any snapshot on the source container which shouldn't exist, then "zfs destroy" those by hand.
